### PR TITLE
Refactor changes

### PR DESCRIPTION
**Pull request by PCSwingle**
Refactor CodeChange into AbstractChange and ConcreteChange as outlined in #91 

Edit: After talking with @biobootloader, we've decided to change the structure quite a bit. This is now the general layout:

Model output -> Parser (this parser can be injected in, so any parser will work) -> new FileEdit format that groups by file and only has 'Replacement' subedits -> User filter (make these changes, keep these ones, don't use these ones, etc.) -> Conflict Resolution (handled by FileEdit; no conflict resoution needed on a per-parser basis). The final list of FileEdits can be converted into a final file's code lines, which is written to the file by the CodeFileManager.

Instead of completely changing the Parser and CodeChange files, I've extracted them into OriginalFormatParser and OriginalFormatChange classes; while we don't necessarily need intermediate CodeChanges for every parser, in this case it helps us use our previous parsing code instead of re-writing it all.

TODO:
- [x] Make Parser interface and make it easily injectable
- [x] Add prompts connected to Parser
- [x] Add tests for FileEdit, especially conflict resolution
- [x] Fix all TODO's scattered around the code
- [x] Restructure tests between format specific tests and non-format specific tests

### CODE REVIEW

File: mentat/code_changes/abstract/abstract_change.py
Comment (biobootloader) [inline mentat/code_changes/abstract/abstract_change.py:36]: I don't love using `None` for deletion inside `Rename`. I might prefer explicit `DeleteFile`, `CreateFile`, `RenameFile` classes. Or maybe the name could be `FileUpdate` or something like that instead of `Rename`

Comment (PCSwingle) [reply mentat/code_changes/abstract/abstract_change.py:36]: Hmm, looking back at it I think I agree that having DeleteFile CreateFile and RenameFile classes would be nicer and simpler; I'll switch over to that.

Comment (PCSwingle) [reply mentat/code_changes/abstract/abstract_change.py:36]: Hmm, but thinking about it more I do like having just one class

File: mentat/parsers/block_parser.py
Comment (jakethekoenig) [inline mentat/parsers/block_parser.py:12]: I'd define the string here instead of importing. Or do we plan to keep all the prompts in the same place? I guess I could see that being advantageous if some of us become more specialized "prompt engineers".

Similarly I don't see the value of importing `original_format_parsing` code instead of deleting that file and moving its code here.

File: mentat/parsers/block_parser.py
Comment (jakethekoenig) [inline mentat/parsers/block_parser.py:15]: It's called block because the changes are given in "blocks" like this?
```
@@start
{
    "file": "core/script.py",
    "action": "insert",
    "insert-after-line": 3,
    "insert-before-line": 4
}
@@code
    if name == "Bob":
        print("Nice to see you again!")
@@end
```

**Review by jakethekoenig — COMMENTED** [review]
I really like the direction here. Thanks!

File: mentat/parsers/block_parser.py
Comment (PCSwingle) [reply mentat/parsers/block_parser.py:12]: Hmm, I personally like having the prompt in a separate file; in fact, I would prefer to eventually just have it be a text file (instead of messing around with all of this python dedent stuff and strip for the prompt), and there would be a different text file for each prompt. Definitely open to suggestions on that too!

As for the original_format_parsing (and original_format_change) files, we could definitely move their code to this file; my reasoning for leaving it in those 2 files was that it's more like 'legacy' code; ideally, we don't have to touch that streaming/parsing code, and future streaming parsers we build around the current system will be cleaner (since they were built for this system). I definitely don't have anything against moving both files code to this file though!

File: mentat/parsers/block_parser.py
Comment (PCSwingle) [reply mentat/parsers/block_parser.py:15]: Haha that is the reasoning; I don't like the name but I spent too long trying to think of a better name! Definitely let me know if you have any better ideas!

Comment (jakethekoenig) [reply mentat/parsers/block_parser.py:15]: The name is fine. I was worried future formats could also be reasonably called "block" but thinking about it it seems like a lot of future formats we will try will be easier to name e.g. GitDiffParser, FullFileParser

File: mentat/conversation.py
Comment (PCSwingle) [reply mentat/conversation.py:34]: :+1:

**Comment by PCSwingle** [comment]
This PR is almost ready to merge; I'm going to wait on #101 to merge and then I'll resolve conflicts.

File: mentat/parsers/block_parser.py
Comment (jakethekoenig) [reply mentat/parsers/block_parser.py:12]: I could definitely see putting prompts in plain text. Though the ergonomics of importing a string are maybe a little better than reading a file.

I'm fine with leaving it separated and not cleaning it up to fit more organically in the current system. I imagine after a period of experimentation we'll settle on just one parser that we like best and delete the others from the code base. I could be wrong though. Possibly different models/languages or types of requests will work better with certain formats.

**Comment by PCSwingle** [comment]
> > This PR is almost ready to merge; I'm going to wait on #101 to merge and then I'll resolve conflicts.
> 
> Are the todos still in `file_edit` for a future PR?

The Replacement 'owner' TODO yes; the current parsing format has a one to one 'change' to Replacement ratio, but eventually we might have a format that creates 2 replacements that we want to represent as one 'change' that the user can filter. It isn't a big addition though and just complicates a few things, so I figured we might as well add that when we need it. As for the conflict resolution TODO's, I personally don't think user input for resolving them is super important, but if you or @biobootloader think we should get that done before merging this PR I'm 100% ok with doing it now.

File: pyrightconfig.json
Comment (granawkins) [inline pyrightconfig.json:1]: Nice. was wondering about this.

File: mentat/parsers/block_parser.py
Comment (waydegg) [inline mentat/parsers/block_parser.py:1]: What's your thinking on which classes should be wrapped with `attrs`?

**Review by waydegg — CHANGES_REQUESTED** [review]

**Review by waydegg — COMMENTED** [review]
Left some comments!

File: mentat/parsers/block_parser.py
Comment (biobootloader) [reply mentat/parsers/block_parser.py:12]: I like having the prompts in other files as well. Right now the imported `block_parser_prompt` is our entire original system prompt. Before adding a second output format we'll have to split that into parts: 1) a generic part that explains that the model is part of a code editing system, that it might help the user write code or answer questions, etc. and 2) the output format specific part that gives a description of the response format and examples.

When we make that split lets think about the file format and explore the options @waydegg suggested

Comment (biobootloader) [reply mentat/parsers/block_parser.py:12]: > my reasoning for leaving it in those 2 files was that it's more like 'legacy' code; ideally, we don't have to touch that streaming/parsing code, and future streaming parsers we build around the current system will be cleaner

I'm ok leaving that code there since our intention is we find better output formats and then delete those files. I wouldn't want it hanging around as legacy code for very long though - if we find we still want this output format for something, we'd want to clean it up / integrate it more.

**Comment by biobootloader** [comment]
Thanks @PCSwingle for this important and complicated refactor! And thanks everyone for the quick reviews 🚀 

I'll take a closer look later today after you've had a chance to address comments, let me know when you're ready for that!

I'm excited to try some new output formats!

File: mentat/code_file_manager.py
Comment (PCSwingle) [reply mentat/code_file_manager.py:179]: Hmm, I'm not entirely sure about this because that stages the rename; I don't think we want mentat to stage anything unless the user specifically asks us to; and of course when the user eventually stages the changes git will recognize the rename either way. It is nice having git recognize it immediately though. What do you think?

File: mentat/conversation.py
Comment (PCSwingle) [reply mentat/conversation.py:7]: Ah, good catch! We can talk about changing our lint settings to automatically do whatever we decide next time we meet.

File: mentat/parsers/file_edit.py
Comment (PCSwingle) [reply mentat/parsers/file_edit.py:21]: I agree, it would be good to clean up our code so it isn't so interconnected; we might be able to get it done after refactoring the interface and our singletons.

File: mentat/parsers/file_edit.py
Comment (PCSwingle) [reply mentat/parsers/file_edit.py:24]: When we enter interactive mode we want to be able to veto (and in the future ask for fixes to) specific changes that the model made; right now, we just veto individual Replacements, but if the Replacements had an 'owner' which represents that multiple replacements were made by one vetoable change, we would group them together in interactive mode.

File: mentat/parsers/block_parser.py
Comment (PCSwingle) [reply mentat/parsers/block_parser.py:1]: If it's a dataclass; i.e., meant to pretty much be a struct (although we can add functions them) so that attrs automatically gives us slots, eq, comparisons, and other stuff.

File: mentat/parsers/block_parser.py
Comment (PCSwingle) [reply mentat/parsers/block_parser.py:15]: Hmm, I'm not sure I understand what the File means in this context? Most likely all of our formats will need the model to specify the file.

File: mentat/code_file_manager.py
Comment (biobootloader) [reply mentat/code_file_manager.py:179]: hmm yeah we don't want to stage. I believe `git mv` forces git to recognize it as a rename (which it sometimes fails, if similarity isn't high enough). Let's leave as is for now so we don't stage. Perhaps in the future we could have Mentat "remember" a rename and so when it commits for the user it'll inform git

File: mentat/code_file_manager.py
Comment (biobootloader) [inline mentat/code_file_manager.py:133]: We used to do these checks earlier, right? like when the change was created by parsing?

On the one hand, some these should be rare errors (creating a file that already exists, editing non-existent file). But editing a file not in context might be common, since we are starting to include parts of files in context. It's not ideal if Mentat crashes *after* the user has seen the changes and accepted them / spent time doing interactive selection or providing feedback.

Comment (PCSwingle) [reply mentat/code_file_manager.py:133]: We still do do them in the current parser; but of course, we don't want to copy that code to every parser we make. I agree we should always do it in the parser though; what if we make it so that FileEdit does these checks on creation? That way it would be done at parsing for every parser. This also made me realize that we don't do any rename file checks here anymore; I'll fix that.

Comment (biobootloader) [reply mentat/code_file_manager.py:133]: Yes doing it on FileEdit creation sounds great!

File: mentat/parsers/block_parser.py
Comment (biobootloader) [reply mentat/parsers/block_parser.py:15]: `OriginalFormatParser` would be unambiguous, but `BlockParser` seems fine for now.

The name really just matters if we end up keeping this format around, which we'd only do if it is the best for some models / scenarios. I don't expect that but if it does happen we'll find a name that distinguishes it well from the other formats

File: mentat/parsers/file_edit.py
Comment (biobootloader) [inline mentat/parsers/file_edit.py:179]: Let's print something here so the user knows what happened. Perhaps just say "these two changes conflicted, they've been auto-merged back to back, double check: {the changes back to back}". From the user's perspective they may have seen each change individually and thought everything looked good.

File: mentat/parsers/file_edit.py
Comment (biobootloader) [inline mentat/parsers/file_edit.py:172]: so what happens in that scenario now? does Mentat crash? or does it just randomly put one first?

Comment (PCSwingle) [reply mentat/parsers/file_edit.py:172]: Right now it just puts the insertion first in the list first and the next insertion last

File: mentat/parsers/file_edit.py
Comment (biobootloader) [reply mentat/parsers/file_edit.py:181]: `get_updated_file_lines` ?

File: mentat/parsers/file_edit.py
Comment (biobootloader) [reply mentat/parsers/file_edit.py:172]: how about we treat that scenario the same as the other conflict scenario then, and alert / display to the user what's happening

File: mentat/parsers/block_parser.py
Comment (waydegg) [reply mentat/parsers/block_parser.py:15]: > Hmm, I'm not sure I understand what the File means in this context?

"File" would mean a literal file on the file system. Technically GitDiffParser would be pulling from a file too :/

**Review by biobootloader — APPROVED** [review]
Nice stuff! Feel free to merge after addressing comments

**Pull request merged**
